File: leetcode/problems/intersection_of_two_arrays.rb
# Given two arrays, write a function to compute their intersection.
#
# Example:
# Given nums1 = [1, 2, 2, 1], nums2 = [2, 2], return [2].
#
# Note:
# Each element in the result must be unique.
# The result can be in any order.

# param {Integer[]} nums1
# param {Integer[]} nums2
# return {Integer[]}

def intersection(nums1, nums2)
  result = []

  nums1.sort!
  nums2.sort!

  p1, p2 = 0, 0

  while p1 < nums1.length && p2 < nums2.length
    if nums1[p1] == nums2[p2] && result[-1] != nums1[p1]
      result << nums1[p1]
      p1 += 1
      p2 += 1
    elsif nums1[p1] < nums2[p2]
      p1 += 1
    else
      p2 += 1
    end
  end

  result
end

# Time Complexity - O(nlogn)
# Space Complexity - O(n)

File: leetcode/javascript_problems/intersection_of_two_arrays.js
// Given two arrays, write a function to compute their intersection.
//
// Example:
// Given nums1 = [1, 2, 2, 1], nums2 = [2, 2], return [2].
//
// Note:
// Each element in the result must be unique.
// The result can be in any order.

/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */

var intersection = (nums1, nums2) => {
  let result = new Set();

  nums1.sort((a, b) => a - b);
  nums2.sort((a, b) => a - b);

  let n1 = 0;
  let n2 = 0;

  while (n1 < nums1.length && n2 < nums2.length) {
    if (nums1[n1] === nums2[n2]) {
      result.add(nums1[n1]);
      n1 ++;
      n2 ++;
    } else if (nums1[n1] < nums2[n2]) {
      n1 ++;
    } else {
      n2 ++;
    }
  }

  return Array.from(result);
};
